### PR TITLE
Do not prune nested type

### PR DESCRIPTION
**Pull request by lkerford**
Currently we are pruning protos which provided nested types if no other types in the file is used.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Hm what's the motivation here? The existing behavior seems correct. While nesting implies a relationship, that relationship should not be the sole reason a nested type is kept. It should have an actual path to a prune root through the type system.

**Comment by jrodbx** [comment]
Agree with Jake here.  What's the use case here @lkerford ?

**Comment by lkerford** [comment]
@JakeWharton @jrodbx  we have found cases were we get NPE when compiling protos.

 If protos A depends on a nested type in proto B but none of the none nested type. We remove proto B from the schema. We shouldn't be removing proto B from the schema in this case as it is use. This currently result in NPE mentioned above.

File: wire-schema/src/jvmMain/java/com/squareup/wire/schema/ProtoFile.kt [hunk 1]
```diff
@@ -66,6 +66,19 @@ actual class ProtoFile private constructor(
...
   }
 
+  /**
+   * Returns all types and subtypes which are found in the proto file.
+   */
+  fun typesAndNestedTypes(): List<Type> {
+    val typesAndNestedTypes = mutableListOf<Type>()
+    typesAndNestedTypes.addAll(types)
+    for (type in types) {
+      typesAndNestedTypes.addAll(type.nestedTypes)
```
Comment (swankjesse) [inline wire-schema/src/jvmMain/java/com/squareup/wire/schema/ProtoFile.kt]: change this to be recursive?

File: wire-compiler/src/test/java/com/squareup/wire/schema/WireRunTest.kt [hunk 2]
```diff
@@ -558,6 +588,11 @@ class WireRunTest {
           |package squareup.polygons;
           |message Triangle {
           |  repeated double angles = 1;
+          |   enum Type {
```
Comment (oldergod) [inline wire-compiler/src/test/java/com/squareup/wire/schema/WireRunTest.kt]: nittty as hell but there are 1 too many space on the indention

Comment (lkerford) [reply wire-compiler/src/test/java/com/squareup/wire/schema/WireRunTest.kt]: good catch, will fix

**Pull request merged**